x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/TimeSeriesGeoLineBuckets.java

```diff
@@ -0,0 +1,328 @@
+        Function<Long, Long> circuitBreaker
+    ) {
+        this.valuesSources = valuesSources;
+        this.bucketSize = bucketSize;
+        this.geoLineBuilder = geoLineBuilder;
+        this.currentBucket = NO_BUCKET;
+        this.simplifier = new TimeSeriesGeoLineBuckets.Simplifier(bucketSize, circuitBreaker);
+    }
+
+    /**
+     * Produce the leaf collector for collecting the timestamp (sort field) and geo_point for each document.
+     * Each lucene segment will have its own leaf collector, but it is still possible to have multiple buckets per-leaf.
+     */
+    Leaf forLeaf(AggregationExecutionContext aggCtx) throws IOException {
+        return new Leaf(aggCtx.getLeafReaderContext());
+    }
+
+    void doPostCollection() {
+        // Ensure any last bucket is completed
+        flushBucket(NO_BUCKET);
+    }
+
+    void doClose() {
+        simplifier.reset();
+    }
+
+    /**
+     * Since there is a different instance of the leaf collector for each segment, and within a segment
+     * the bucket id can change (or tsid can change), we need to copy the streaming geometry simplifier memory
+     * into an InternalGeoLine object before collecting the next bucket. This is handled by the Leaf.collect(doc,bucket)
+     * method calling flushBucket(bucket) and any change in bucket id will copy the memory and prepare the simplifier
+     * for the next bucket.
+     */
+    private void flushBucket(long bucket) {
+        if (bucket != currentBucket) {
+            if (currentBucket != NO_BUCKET) {
+                if (geoLines.containsKey(currentBucket)) {
+                    throw new IllegalStateException("Geoline already exists for bucket " + currentBucket);
+                }
+                geoLines.put(currentBucket, geoLineBuilder.apply(currentBucket));
+                simplifier.reset();
+            }
+            currentBucket = bucket;
+        }
+    }
+
+    /**
+     * Build the aggregation based on saved state from the collector phase.
+     * Note that the collector phase does not know about sort-order, so if the order is not the native time-series
+     * order of 'descending', this is where we can do an efficient in-place reverse ordering.
+     * Also note that the collector phase will re-use the same memory (in the StreamingGeometrySimplifier) for each bucket,
+     * so it is important that this method is called to copy that 'scratch' memory into the final InternalGeoLine aggregation
+     * objects between each bucket collection phase.
+     */
+    InternalGeoLine buildAggregation(long bucket, String name, Map<String, Object> metadata, boolean includeSorts, SortOrder sortOrder) {
+        LineStream line = simplifier.produce();
+        boolean complete = simplifier.length() < bucketSize;
+        if (line.isEmpty()) {
+            // TODO: Perhaps we should return an empty geo_line here?
+            throw new IllegalStateException("Invalid bucket " + bucket + " for geo_line");
```

Comment:
yes, I think we should return an empty aggregation instead of throwing an error

Reply:
Done